xonsh/base_shell.py

```diff
@@ -343,11 +343,14 @@ def _append_history(self, tee_out=None, **info):
         information is available.
         """
         hist = builtins.__xonsh_history__  # pylint: disable=no-member
+        env = builtins.__xonsh_env__
         info['rtn'] = hist.last_cmd_rtn
         tee_out = tee_out or None
         last_out = hist.last_cmd_out or None
         if last_out is None and tee_out is None:
             pass
+        elif not env.get('XONSH_STORE_STDOUT'):
+            pass
```

Comment:
Well, I asked the quite same question here: https://github.com/xonsh/xonsh/pull/1970#discussion_r90779531

We are using `__xonsh_history__.rtns[-1]` as the counterpart of `$?` in Bash. Currently for json based history, if we set `ignoredups`, we will have an issue with it:

```
last-rtn[<not-set>]$ ls foo  # foo not exist, rtn is 1

# in another xonsh session, we create the file `foo` out.

last-rtn[1]$ ls foo  # foo now exists, rtn 0, but __xonsh_history__.rtns[-1] not got updated.
last-rtn[1]$ 
```
This is an issue because I think xonsh users usually use `__xonsh_history__.rtns[-1]` with colors in prompt string to know whether last cmd is succeed or not.